lib/cassandra/column_container.rb

```diff
@@ -0,0 +1,341 @@
+    attr_reader :clustering_columns
+    # @return [Array<Cassandra::Column>] primary key of this column-container. It's the combination of `partition_key` and `clustering_columns`.
+    # @note This composition produces a flat list, so it will not be possible for the caller to distinguish partition-key columns from clustering-columns.
+    attr_reader :primary_key
+
+    # @private
+    def initialize(keyspace,
+                   name,
+                   partition_key,
+                   clustering_columns,
+                   other_columns,
+                   options,
+                   id)
+      @keyspace           = keyspace
+      @name               = name.freeze
+      @partition_key      = partition_key.freeze
+      @clustering_columns = clustering_columns.freeze
+      @options            = options
+      @id                 = id
+
+      # Make one array of all the columns, ordered with partition key, clustering
+      # columns, then other columns. Make a hash as well, to support random access
+      # to column metadata for a given column name. Save off the primary key (which
+      # is partition-key + clustering-columns) while we're at it.
+
+      @primary_key = @partition_key.dup.concat(@clustering_columns).freeze
+      @columns = @partition_key.dup.concat(other_columns).freeze
+      @columns_hash = @columns.each_with_object({}) do |col, h|
+        h[col.name] = col
+      end
+    end
+
+    # @param name [String] column name
+    # @return [Boolean] whether this column-container has a given column
+    def has_column?(name)
+      @columns_hash.key?(name)
+    end
+
+    # @param name [String] column name
+    # @return [Cassandra::Column, nil] a column or nil
+    def column(name)
+      @columns_hash[name]
+    end
+
+    # Yield or enumerate each column defined in this column-container
+    # @overload each_column
+    #   @yieldparam column [Cassandra::Column] current column
+    #   @return [Cassandra::ColumnContainer] self
+    # @overload each_column
+    #   @return [Array<Cassandra::Column>] a list of columns
+    def each_column(&block)
+      if block_given?
+        @columns.each(&block)
+        self
+      else
+        @columns
+      end
+    end
+    alias columns each_column
+
+    # @private
+    # keyspace may be nil because when this object was constructed, we didn't have
+    # its keyspace constructed yet. So allow updating keyspace if it's nil, thus
+    # allowing fetchers to create keyspace, table/view, and hook them together without
+    # worrying about chickens and eggs.
+    # NOTE: Ignore the set request if the keyspace attribute is already set.
+    def set_keyspace(keyspace)
+      @keyspace = keyspace.freeze unless @keyspace
+    end
+
+    # @private
+    def inspect
+      "#<#{self.class.name}:0x#{object_id.to_s(16)} " \
+          "@keyspace=#{@keyspace} @name=#{@name}>"
+    end
+
+    # @private
+    def eql?(other)
```

Comment:
Good call on inspect; will remove. eql? is invoked by the subclasses' eql?'s to compare the ColumnContainer-owned fields.